rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,480 @@
+
+Another possibility is that the Kube Agent might use the invite token each time the pod is recycled, meaning that it issues a join request to the cluster every time it starts and receives a new identity from the Auth service. This means that the invite token must be a static/long-lived token, otherwise after a while the agent could not register himself in the cluster because the invite token expired. This approach is not recommended and might cause security flaws because the join token can be stolen or guessed.
+
+One solution that might address all the issues referenced above is to allow the Kube agent to use Kubernetes secrets as storage backend. This allows not only that the agent is able to run stateless depending only on native objects generally available in any Kubernetes cluster, but also that the agent might support dynamic short-lived invite tokens with no dependency on external storage.
+
+## Details
+
+### Limitations
+
+High availability can only be achieved using Statefulsets and not deployments. This means that if the desired number of replicas is bigger than one, i.e, `replicas > 1`, the [Helm chart](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-kube-agent) has to switch to Statefulset objects instead of Deployments. This change is required because some Kubernetes invariant must be kept between restarts in order to correctly map each agent pod to its identity stored in the secret. The invariant used is the Statefulset pod names `{{ .Release.Name }}-{0...replicas}}`.
+
+Given this, it is required to expose the `$TELEPORT_REPLICA_NAME` environment variable to each pod in order to the backend storage be able to write the identity separately. The values for `$TELEPORT_REPLICA_NAME` are dependent on the object type:
+
+- Deployment: `TELEPORT_REPLICA_NAME` is a constant string `{{ .Release.Name }}`.
+- Statefulset: `TELEPORT_REPLICA_NAME` is a dynamic value provided by Kubernetes `fieldPath: metadata.name`.
+
+When an operator wants to upgrade from an old chart to the Helm chart that implements this RFD and his setup has no storage, it's running as Deployment, and the replica number was bigger than 1, it will lead into a switch from Deployment to a Statefulset. 
+This change is managed by Helm by destroying the Deployment object and creating the new Statefulset. During this transition, even if the operator has the PodDisruptionBudget object enabled, the Kubernetes cluster might become inaccessible from Teleport because there is no guarantee that the system has at least one agent replica running.
+
+Another limitation that appears when the operator wants to change the number of replicas is that this process must necessarily be executed through Helm. If the operator decides to do a manual increase of replicas by editing the Kubernetes objects, it will have the following limitations, depending on the object type:
```

Comment:
@r0mant is worth trying to look into a zero downtime process or can we just advise the users that require no downtime to install teleport agent with a different release name and later delete the previous one?